### ISSUE TITLE
Suggested code from omit_local_variable_types linter fails at runtime

### ISSUE THREAD

**Issue by TheSpydog** [dart-lang/linter#3469]

**Comment by eernstg** [comment]
> I believe the recommended style is to use var and move the type information to the initializer expression

That recommendation has indeed been put forward several times.

However, we might note that this is to some extent a recommendation to avoid type inference, because it requires the developer to provide feed-forward type arguments in various locations in the initializing expression, whereas the variable with an explicit type annotation (that is `SomeType<Maybe, With, Arguments> x = e;`) will declaratively state the desired result, and leave it up to the type inferencer to compute ("backwards") which type arguments in the initializing expression are needed in order to reach that goal.

I usually push the idea that `omit_local_variable_types` should only flag declarations where the initializing expression has an "obvious" type, which is basically the same thing as saying: If you are using non-trivial type inference then it's OK to declare the type.

**Comment by munificent** [comment]
> I believe the recommended style is to use `var` and move the type information to the initializer expression. @munificent can confirm or refute that premise.

That's right, and [the "Effective Dart" rule for it](https://dart.dev/guides/language/effective-dart/design#dont-redundantly-type-annotate-initialized-local-variables) links to this lint.

> I usually push the idea that `omit_local_variable_types` should only flag declarations where the initializing expression has an "obvious" type, which is basically the same thing as saying: If you are using non-trivial type inference then it's OK to declare the type.

That's what we say for [fields and top-level variables](https://dart.dev/guides/language/effective-dart/design#do-type-annotate-fields-and-top-level-variables-if-the-type-isnt-obvious). For locals, we felt it was better to have a simple blanket rule that can be applied mechanically. Since a local variable never bleeds out into a library's exported API, pinning down the precise type is less important.

**Comment by eernstg** [comment]
@munificent wrote:
> Since a local variable never bleeds out into a library's exported API, pinning down the precise type is less important

Of course, all these considerations are not relevant to `var x = C<Some, Type, Arguments>(...);` where `C` is a class, because we can easily see that the type of `x` is `C<Some, Type, Arguments>`. But `omit_local_variable_types` should flag such cases exactly because the type is obvious.

Otherwise, the reason why I keep pushing for an `omit_local_variable_types` that doesn't forbid helpful declared types is not that anyone a hundred miles away will see and depend on the declared type of the local variable&mdash;as you mention, that can't happen anyway.

My main reason is that the placement of the type information as a declared type is a _declarative_ approach that relies on type inference to obtain the actual type arguments at various locations in the initializing expression. So we specify the desired end result, and ask the computer to sort out all the details in order to reach that result.

In contrast, the placement of one or more actual type arguments in the initializing expression requires the developer to (1) sort out all those details themselves (in order to know which type arguments to provide, and where), and (2) redo the information propagation in order to compute the actual declared type of the variable whenever they need to read and understand the code.

In general, I'd consider "state the desired end result and let the computer sort out the details" to be a more rational and useful approach than "sort out the details manually, write some information that will produce the end result after any number of complex steps that aren't visible in the source code, and ask the developers who will ever need to read and understand this code to redo those steps manually".

**Comment by munificent** [comment]
> My main reason is that the placement of the type information as a declared type is a _declarative_ approach that relies on type inference to obtain the actual type arguments at various locations in the initializing expression. So we specify the desired end result, and ask the computer to sort out all the details in order to reach that result.

I definitely see the appeal of that as a consistent approach. Basically, there's two options to avoid users redundantly specifying information:

1. They specify it in types and we push that over into expressions. (What you suggest here.)
2. They specify it in expressions and we push that over into types. (What Effective Dart and this lint suggest.)

For top-level variables and fields we encourage users to do (1) because we *want* them to think about the type signatures of their API explicitly since a change to those types can have far-reaching effects.

For local variables, we lean towards (2) because expressions carry type information that can be inferred from *values*. In:

```dart
var x = [1];
```

They don't have to write `List` or `int` at all because it can be inferred entirely from the syntax of the list literal and the type of the integer literal inside it. So (2) ends up being maximally terse for most cases—no type arguments needed to be written at all, either in the variable's type annotation *or* the initializing expression.

Users really like this kind of inference, so we want them to use it when they can. But some initializing expressions don't carry enough information to infer all type arguments:

```dart
var x = [];
```

That leaves us with two options:

```dart
var x = <int>[];
List<int> x = [];
```

We went with the former because that leads to more consistent code:

```dart
var x = [1];
var y = <int>[];
var z = [3];
```

Versus:

```dart
var x = [1];
List<int> y;
var z = [3];
```

In the latter example, `y` sticks out as being different when really it's just another local variable.

There are still times when you need to annotate a local variable, if the inferred type is not what you want or there is no initializer. But those are rare and stand out deliberately.

Philosophically, I think there's an argument to prefer specifying information in expressions and inferring types from them because expressions are usually more precise than types since they describe an actual value.
